File: mercadolivre/src/main/java/com/jornadadev/mercadolivre/entity/Usuario.java
package com.jornadadev.mercadolivre.entity;

import lombok.Getter;
import lombok.ToString;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.util.Assert;
import org.springframework.util.StringUtils;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.validation.constraints.Email;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.PastOrPresent;
import javax.validation.constraints.Size;
import java.time.LocalDateTime;

@Entity
@ToString
@Getter
public class Usuario {
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    private Long id;

    @Email
    @NotEmpty
    private String email;

    @NotEmpty
    @Size(min = 6)
    private String senha;

    @NotNull
    @PastOrPresent
    private LocalDateTime dataCadastro;

    /**
     * @param email string no formato de email
     * @param senha string em texto limpo
     */
    public Usuario(String email, String senha) {
        Assert.isTrue(StringUtils.hasLength(email), "Email não pode ser em branco");
        Assert.isTrue(StringUtils.hasLength(senha), "Senha não pode ser em branco");
        Assert.isTrue(senha.length() >= 6, "Senha deve ter no mínimo 6 caracteres.");
        this.email = email;
        this.senha = new BCryptPasswordEncoder().encode(senha);
        this.dataCadastro = LocalDateTime.now();
    }

    @Deprecated
    public Usuario(){

    }
}

File: yfood/src/main/kotlin/com/jornadadev/yfood/entities/Usuario.kt
package com.jornadadev.yfood.entities

import com.jornadadev.yfood.listarformaspagamento.regrasfraude.RegraFraude
import org.springframework.util.Assert
import javax.persistence.*
import javax.validation.constraints.Email
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

@Entity
data class Usuario(
        @NotBlank
        @NotNull
        @Email
        val email: String,
        @Size(min = 1)
        @ElementCollection
        @Enumerated(value = EnumType.STRING)
        val formaPagamentoEnum: Set<FormaPagamentoEnum>

) {
    @OneToMany(mappedBy = "usuario", cascade = [CascadeType.MERGE])
    private val selecoes: MutableList<RestauranteSelecionado> = mutableListOf()
    init {
        println("Criando objeto Usuario")
        Assert.hasText(email, "Email não pode ser branco.")
        Assert.isTrue(formaPagamentoEnum.isNotEmpty(), "Tipos de pagamento não pode ser vazio.")
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0

    fun filtraFormasPagamento(restaurante: Restaurante, regrasFraude: Set<RegraFraude>) : Set<FormaPagamentoEnum> {
        return this.formaPagamentoEnum
                .filter { restaurante.aceita(it) }
                .filter { formaPagamento ->
                    regrasFraude.all {
                        it.aceita(this, formaPagamento)
                    }
                }
                .toSet()
    }

    fun podePagar(
            restaurante: Restaurante,
            formaPagamento: FormaPagamentoEnum,
            regrasFraude: Set<RegraFraude>
    ) : Boolean = filtraFormasPagamento(restaurante, regrasFraude).contains(formaPagamento)

    fun registraSelecao(restaurante: Restaurante) {
        selecoes.add(RestauranteSelecionado(usuario = this, restaurante = restaurante))
    }

    fun selecionou(restaurante: Restaurante, nVezes: Int): Boolean {
        val count = this.selecoes.count {
            it.restaurante == restaurante
        }
        return count >= nVezes
    }


}
